webonary-cloud-api/tools/flexXhtmlParser.ts

```diff
@@ -0,0 +1,187 @@
+      : '';
+    /*
+    <span class="lexemeform">
+    <span>
+        <audio id="g636928554709180064rɩɩb_bʋkr_so-tũudga">
+            <source src="AudioVisual\\636928554709180064rɩɩb bʋkr so-tũudga.mp3"/>
+        </audio>
+        <a class="mos-Zxxx-x-audio" href="#g636928554709180064rɩɩb_bʋkr_so-tũudga" onclick="document.getElementById('g636928554709180064rɩɩb_bʋkr_so-tũudga').play()"/>
+    </span>
+    </span>
+    */
+
+    const audioId = $('span.lexemeform span audio').attr('id') ?? '';
+    const audioSrc =
+      $('span.lexemeform span audio source')
+        .attr('src')
+        ?.replace(/\\/g, '/') ?? '';
+    const fileClass = $('span.lexemeform span a').attr('class') ?? '';
+    const audio: EntryFile = { id: audioId, src: audioSrc, fileClass };
+
+    /* 
+    <span class="pictures">
+    <div class="picture">
+        <img class="thumbnail" id="g9f0ee134-8268-45b1-ae6d-4bb7aac97920" src="pictures\tube_digestif.jpg"/>
+        <div class="captionContent">
+            <span class="headword">
+                <span lang="mos">rɩɩb bʋkr so-tũudga</span>
+            </span>
+        </div>
+    </div>
+    </span>
+    */
+
+    const pictures: EntryFile[] = [];
+    $('span.pictures div.picture img').map((i, elem) => {
+      const id = $(elem).attr('id') ?? '';
+      const src =
+        $(elem)
+          .attr('src')
+          ?.replace(/\\/g, '/') ?? '';
+      const caption =
+        $(elem)
+          .next()
+          .find('span.headword span')
+          .text() ?? '';
+      pictures.push({ id, src, caption });
+    });
+
+    const pronunciations: EntryValue[] = [];
+    $('span.pronunciations span.pronunciation span span').map((i, elem) => {
+      const type =
+        $(elem)
+          .parent()
+          .attr('class') ?? '';
+      const lang = $(elem).attr('lang') ?? '';
+      const value = $(elem).text();
+      pronunciations.push({ lang, value, type });
+    });
+
+    const partOfSpeech: EntryValue = {
+      lang:
+        $(
+          'span.senses span.sharedgrammaticalinfo span.morphosyntaxanalysis span.partofspeech span',
+        ).attr('lang') ?? '',
+      value: $(
+        'span.senses span.sharedgrammaticalinfo span.morphosyntaxanalysis span.partofspeech span',
+      ).text(),
+    };
+
+    const definitionOrGloss: EntryValue[] = [];
+    $('span.senses span.sensecontent span.sense span.definitionorgloss span').map((i, elem) => {
+      const lang = $(elem).attr('lang');
+      const value = $(elem).text();
+      if (lang && value) {
+        // TODO: this was in moore, but necessary?
+        // if ($(elem).prev().hasClass('writingsystemprefix')) {
+        definitionOrGloss.push({ lang, value });
+      }
+    });
+
+    const reverseLetterHeads = definitionOrGloss.reduce(
+      (letterHeads: EntryValue[], entry: EntryValue): EntryValue[] => {
+        const { lang } = entry;
+        const value = entry.value.toLowerCase().substring(0, 1);
```

Comment:
This might be good enough to make it to production. I thought we would have a problem with Spanish digraphs (I have seen people do this in their Spanish reversals in Flex, but the current recommendation for Spanish does not treat digraphs as separate indexes in a dictionary. So all the most common analysis languages should work.

Reply:
Yes, perhaps. I was planning to do more to make sure collations are honored. Will need more research on FLex, Webonary, and Mongo. Will keep in mind :-)